engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java

```diff
@@ -1461,6 +1460,7 @@ else if ( propertyPathNode.getKey() != null ) {
 
 		metaConstraints.addAll( propertyMetaData.getConstraints() );
 		typeArgumentConstraints.addAll( propertyMetaData.getTypeArgumentsConstraints() );
+		propertyPath.removeLeafNode();
```

Comment:
Not sure I understand the rationale of this commit. Maybe a more explicit commit message and a comment around the two `propertyPath.removeLeafNode()` would help understand what is going on?